### PR TITLE
[FLINK-12541][container][python] Add support for Python jobs in build script

### PR DESCRIPTION
**Pull request by dianfu**
## What is the purpose of the change

*This pull request adds support to build job specific docker image for Python Table API jobs.*

## Brief change log

  - *Improves the build script to support build a Python job specific docker image*

## Verifying this change

Verify this change manually.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by sunjincheng121** [comment]
@flinkbot attention @tillrohrmann

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 1]
```diff
@@ -73,27 +76,31 @@
 			@Nonnull JobID jobId,
 			@Nonnull SavepointRestoreSettings savepointRestoreSettings,
 			@Nonnull String[] programArguments,
-			@Nullable String jobClassName) {
-		this(jobId, savepointRestoreSettings, programArguments, jobClassName, JarsOnClassPath.INSTANCE);
+			@Nullable String jobEntryPointName,
+			@Nullable String jobPythonArtifacts) {
+		this(jobId, savepointRestoreSettings, programArguments, jobEntryPointName, jobPythonArtifacts, JarsOnClassPath.INSTANCE);
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: Why do we need to touch the `ClassPathJobGraphRetriever` at all? This class should not need to know about Python. Otherwise we need to add special casing for all supported languages in the future. This does not seem right.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 2]
```diff
@@ -110,19 +117,38 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
...
 		try {
 			final Class<?> mainClass = getClass().getClassLoader().loadClass(entryClass);
-			return new PackagedProgram(mainClass, programArguments);
-		} catch (ClassNotFoundException | ProgramInvocationException e) {
+			return new PackagedProgram(mainClass, getProgramArguments(isPythonProgram));
+		} catch (ProgramInvocationException e) {
 			throw new FlinkException("Could not load the provided entrypoint class.", e);
+		} catch (ClassNotFoundException e) {
+			if (!isPythonProgram && jobPythonArtifacts != null) {
+				return createPackagedProgram("org.apache.flink.python.client.PythonDriver", true);
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: This is overly complicated and error prone. Why not starting the cluster entrypoint with `o.a.f.python.client.PythonDriver` as `jobClassName`?

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 3]
```diff
@@ -110,19 +117,38 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
...
+
+	private String[] getProgramArguments(final boolean isPythonProgram) throws FlinkException {
+		if (isPythonProgram) {
+			final String[] args = new String[programArguments.length + 4];
+			args[0] = "pym";
+			args[1] = getJobEntryPointNameOrScanClassPath();
+			args[2] = "pyfs";
+			args[3] = jobPythonArtifacts;
+			System.arraycopy(programArguments, 0, args, 4, programArguments.length);
+			return args;
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: Why do we need to do this command line argument magic? Just because `PythonDriver` expects the arguments to be passed in a special order? Why not making the `PythonDriver` more flexible.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java [hunk 4]
```diff
@@ -69,7 +74,12 @@ JobID getJobId() {
...
 
 	@Nullable
-	String getJobClassName() {
-		return jobClassName;
+	String getJobEntryPointName() {
+		return jobEntryPointName;
+	}
+
+	@Nullable
+	String getJobPythonArtifacts() {
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java]: Why is this needed?

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java [hunk 5]
```diff
@@ -45,12 +45,29 @@
...
+
+	private static final Option JOB_ENTRYPOINT_NAME_OPTION = Option.builder("j")
+		.longOpt("job-entrypoint")
+		.required(false)
+		.hasArg(true)
+		.argName("job entrypoint name")
+		.desc("Entrypoint name of the job to run.")
+		.build();
+
+	private static final Option JOB_PYTHON_ARTIFACTS_OPTION = Option.builder("ja")
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: Why does this need to be handled by the `StandaloenJobClusterConfiguration` and cannot be parsed by the `PythonDriver` program?

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java [hunk 6]
```diff
@@ -53,26 +53,31 @@
 	private final String[] programArguments;
 
 	@Nullable
-	private final String jobClassName;
+	private final String jobEntryPointName;
+
+	@Nullable
+	private final String jobPythonArtifacts;
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java]: I think it is not a good idea to add special case logic into the `StandaloneJobClusterEntrypoint`. Why do you think this is needed?

**Comment by tillrohrmann** [comment]
Why has this PR the same Flink issue assigned as #8532? Every PR should have it's own JIRA issue as stated in the [contribution guidelines](https://flink.apache.org/contribute-code.html).

**Comment by sunjincheng121** [comment]
Thanks for the PR @dianfu 
I think the suggestion from @tillrohrmann makes more sense to me. we should let the PR have its own JIRA.
furthermore,  the improvement of `pyfs` also can in a new PR (with a new JIRA).
What do you think?

**Comment by dianfu** [comment]
@tillrohrmann Thanks a lot for your review. 
Your suggestion makes much sense to me. I have created a dedicated JIRA [FLINK-12788](https://issues.apache.org/jira/browse/FLINK-12788) for this PR. Regarding to the changes to `StandaloneJobClusterEntrypoint`, agree that there should not be special logic for Python. I will revert that part of changes.

@sunjincheng121 Thanks a lot for your review. I have created a separate JIRA [FLINK-12787](https://issues.apache.org/jira/browse/FLINK-12787) for the PythonDriver improving. Then we can focus this PR on the build script changes for Python jobs support. 

Will updated the PR later today.

**Comment by dianfu** [comment]
@tillrohrmann @sunjincheng121 I have updated the PR and the changes are only related to the build script for Python jobs support. Looking forward to your feedback.

**Comment by sunjincheng121** [comment]
Thanks for the update @dianfu!

Currently, the PR looks pretty clean form my side. only one improvement I am not pretty sure that is build command option `--opt-jars`.  this way the opt jar does not need upload. but without this option, the user can build a fat jar with option jars. So this change is an improvement(not necessary change) So we may need some opinion from @tillrohrmann!
Otherwise the PR LGTM. I'll merge it when @tillrohrmann says ok!

Best,
Jincheng

**Comment by sunjincheng121** [comment]
I think we can remove the `--opt-jars` due to we @StephanEwen already bring up the discussion for put the table JAR into lib. detail can be found here:  http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Putting-Table-API-jars-in-lib-by-default-td29492.html
What do you think?

**Comment by dianfu** [comment]
@sunjincheng121 Make sense to me as `--opt-jars` is not a must have option. Considering that the table JAR will be put into lib by default in the future, removing this option is good for me as the table JAR is the motivation for this option. Have updated the PR. Looking forward to your response.

File: flink-container/kubernetes/README.md [hunk 7]
```diff
@@ -29,7 +29,7 @@ In non HA mode, you should first start the job cluster service:
 
 In order to deploy the job cluster entrypoint run:
 
-`FLINK_IMAGE_NAME=<IMAGE_NAME> FLINK_JOB=<JOB_NAME> FLINK_JOB_PARALLELISM=<PARALLELISM> envsubst < job-cluster-job.yaml.template | kubectl create -f -`
```
Comment (tillrohrmann) [inline flink-container/kubernetes/README.md]: Why did we remove `FLINK_JOB` here?

File: flink-container/docker/README.md [hunk 8]
```diff
@@ -13,26 +13,28 @@ Install the most recent stable version of [Docker](https://docs.docker.com/insta
 Images are based on the official Java Alpine (OpenJDK 8) image.
 
 Before building the image, one needs to build the user code jars for the job.
-Assume that the job jar is stored under `<PATH_TO_JOB_JAR>` 
+Assume that the job jar is stored under `<COMMA_SEPARATED_PATH_TO_JOB_ARTIFACTS>`
```
Comment (tillrohrmann) [inline flink-container/docker/README.md]: Seems not to fit. I guess the sentence needs to be changed to something like: `A Flink job can consist of multiple jars. In order to specify the required jars, they need to be passed to `--job-artifacts` of the build script. The individual paths are comma separated.

File: flink-container/kubernetes/README.md (hunk 7 above)
Comment (dianfu) [reply flink-container/kubernetes/README.md]: I found that FLINK_JOB has been removed in commit [753e0c6](https://github.com/dianfu/flink/commit/753e0c618a30a19954761cd75fadba4a89836ae1) and just correct the documentation here. What do you think?

File: flink-container/docker/README.md (hunk 8 above)
Comment (dianfu) [reply flink-container/docker/README.md]: +1. Only one concern: what about changing `jars` to `artifacts`?

**Comment by dianfu** [comment]
@tillrohrmann Thanks a lot for your comments. I have updated the PR and addressed one of your concerns. Regarding to FLINK_JOB, this option has been removed and the documentation has not been updated. Just correct the documentation here. Please let me know if you feel that change doesn't make sense or you prefer to address that in another PR.

File: flink-container/kubernetes/README.md (hunk 7 above)
Comment (tillrohrmann) [reply flink-container/kubernetes/README.md]: Good catch. Now this makes sense.

**Comment by knaufk** [comment]
FYI: This change broke `test_docker_embedded_job.sh` and `test_kubernetes_embedded_job.sh` as they were still using the `--job-jar` argument. I have included a `[hotfix]` in https://github.com/apache/flink/pull/874.

**Comment by dianfu** [comment]
@knaufk Thanks a lot for the hotfix. The fix makes sense to me.
